### PR TITLE
point-in-time endpoint for column-level lineage

### PR DESCRIPTION
**Pull request by pawel-big-lebowski**
Signed-off-by: Pawel Leszczynski <leszczynski.pawel@gmail.com>

### Problem

Point in time column lineage endpoint - implementation.

Closes: #2262

### Solution

 * Dataset or job version from a request is used to determine date time when a version was created. 
 * `column_lineage` table is filtered for rows earlier than a determined date time. 
 * Column-lineage rows, dataset rows and dataset_version rows make use of the same date time. 
 * The combination of `point-in-time` and `withDownstream` features does not suit well and deserves a separate PR if needed. 

 
> **Note:** All database schema changes require discussion. Please [link the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) for context.

### Checklist

- [x] You've [signed-off](https://github.com/MarquezProject/marquez/blob/main/CONTRIBUTING.md#sign-your-work) your work
- [x] Your changes are accompanied by tests (_if relevant_)
- [x] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained
- [ ] You've updated any relevant documentation (_if relevant_)
- [x] You've updated the [`CHANGELOG.md`](https://github.com/MarquezProject/marquez/blob/main/CHANGELOG.md#unreleased) with details about your change under the "Unreleased" section (_if relevant, depending on the change, this may not be necessary_)
- [ ] You've versioned your `.sql` database schema migration according to [Flyway's naming convention](https://flywaydb.org/documentation/concepts/migrations#naming) (_if relevant_)
- [ ] You've included a [header](https://github.com/MarquezProject/marquez/blob/main/CONTRIBUTING.md#copyright--license) in any source code files (_if relevant_)

### CODE REVIEW

File: api/src/main/java/marquez/common/models/JobVersionId.java [hunk 1]
```diff
@@ -19,4 +19,9 @@ public class JobVersionId {
   @NonNull NamespaceName namespace;
   @NonNull JobName name;
   @NonNull UUID version;
+
+  public static JobVersionId of(
```
Comment (mobuchowski) [inline api/src/main/java/marquez/common/models/JobVersionId.java:23]: why not `new` if we already have that constructor?

File: api/src/main/java/marquez/service/ColumnLineageService.java [hunk 2]
```diff
@@ -180,4 +242,6 @@ public void enrichWithColumnLineage(List<Dataset> datasets) {
         .filter(dataset -> datasetLineage.containsKey(dataset))
         .forEach(dataset -> dataset.setColumnLineage(datasetLineage.get(dataset)));
   }
+
+  private record ColumnNodes(Instant createdAtUntil, List<UUID> nodeIds) {}
```
Comment (mobuchowski) [inline api/src/main/java/marquez/service/ColumnLineageService.java:246]: `record` 👍

**Review by mobuchowski — APPROVED** [review]

File: api/src/main/java/marquez/common/models/JobVersionId.java (hunk 1 above)
Comment (pawel-big-lebowski) [reply api/src/main/java/marquez/common/models/JobVersionId.java:23]: We do have such a convention for model classes and I kept constant with it.

Comment (wslulciuc) [reply api/src/main/java/marquez/common/models/JobVersionId.java:23]: Yeah, it's a factory method we have for our models as convention throughout the code base. Thanks @pawel-big-lebowski for keeping things constant ❤️

**Pull request merged**